server/src/main/java/org/opensearch/index/mapper/DefaultDerivedFieldResolver.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.mapper;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.common.regex.Regex;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.script.Script;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.opensearch.index.mapper.FieldMapper.IGNORE_MALFORMED_SETTING;
+
+/**
+ * Accepts definition of DerivedField from search request in both forms: map parsed from SearchRequest and {@link DerivedField} defined using client.
+ * The object is initialized per search request and is responsible to resolve {@link DerivedFieldType} given a field name.
+ * It uses {@link FieldTypeInference} to infer field type for a nested field within DerivedField of {@link DerivedFieldSupportedTypes#OBJECT} type.
+ */
+public class DefaultDerivedFieldResolver implements DerivedFieldResolver {
+    private final QueryShardContext queryShardContext;
+    private final Map<String, DerivedFieldType> derivedFieldTypeMap = new ConcurrentHashMap<>();
```

Comment:
I couldn't think of any use case but was hesitant in making it thread unsafe. Same instance of `DerivedFieldResolver` will be used to resolve one or more derived field(s) used in the query from different phases of query separately. I concerned about scenarios, lets say resolution of fields start happening in parallel in future or 2 different phases of query execute in parallel and combine results, again a future possibility. 
I can make it thread unsafe with documentation, what do you think?

Reply:
The reason I'm caching is that the same derived field will be resolved multiple times from different phases of the query. Since inferring is an costly operation, we can avoid re-computation.